### PR TITLE
Add `tfq.math.mps_1d_expectation` for 1D MPS

### LINKED ISSUE #603: [Design] Implement MPS ops under `tfq.math`
With PRs like:
https://github.com/quantumlib/qsim/pull/374
and
https://github.com/quantumlib/qsim/pull/372
going in, It makes sense for us to start looking into upgrading to the latest qsim so that we can add the MPS ops mentioned [here](https://github.com/tensorflow/quantum/issues/380#issuecomment-724557523). In particular I think we should have something like:
1. `tfq.math.1d_expectation`
2. `tfq.math.1d_samples`
3. `tfq.math.1d_sampled_expectation`
With C++ code very similar to what we already have in https://github.com/tensorflow/quantum/blob/master/tensorflow_quantum/core/ops/tfq_simulate_expectation_op.cc (except with additional error checking to ensure the circuit is 1d and using the `MPSSimulator` from qsim etc.)

### PR DESCRIPTION
**Pull request by jaeyoo**
Resolve #603 1st of the three ops.

### CODE REVIEW

**Review by MichaelBroughton — CHANGES_REQUESTED** [review]
This is a great first pass @jaeyoo ! Would you be able to add some tests that verify expectations values computed using cirq's built in expectation calculation functions ? Also removing some of the print statements from the code  :)

File: tensorflow_quantum/core/src/program_resolution.cc [hunk 1]
```diff
@@ -308,4 +308,68 @@ Status ResolveSymbols(
...
+                                               .arg_value()
+                                               .string_value();
+        if (!control_qubits.empty()) {
+          std::vector<absl::string_view> control_ids =
+              absl::StrSplit(control_qubits, ',');
+          num_control_qubits = control_ids.size();
+        }
+        if (num_qubits + num_control_qubits > 2) {
+          return Status(tensorflow::error::INVALID_ARGUMENT,
+                        "An operation contains more than two qubits.");
```
[the remark this answers is not in the document]
Comment (jaeyoo) [reply tensorflow_quantum/core/src/program_resolution.cc:337]: done. thank you for the catch :)

**Comment by MichaelBroughton** [comment]
@jaeyoo any updates on this ? (My workstation has gone offline and won't be into the office for a while to turn it back on).

**Comment by jaeyoo** [comment]
> @jaeyoo any updates on this ? (My workstation has gone offline and won't be into the office for a while to turn it back on).

I am still working on how to fix the `delete` issue...

**Comment by jaeyoo** [comment]
@MichaelBroughton I could fix the error finally. It was not easy because it had multiple reasons.

1. It turns out that we can't use `tfq::QsimFor` : I realized that the current `MPSSimulator` doesn't accept `qsim::SequantialFor`. It accepts `qsim::For` only. On top of it, for initialization, dummy value 1 is used for `ForArgs`, not `tfq_for` value.
2. Even if the theoretical minimum of qsim MPS bond_dim is 2 (as shown [here](https://github.com/quantumlib/qsim/blob/5a7a1319abe95d76115d5ba428d2e9bd3f097d30/lib/mps_statespace.h#L98)), but the real test seems to require minimum 4 (as qsim MPSSimulator test code.), because the weird `double free` error occurred from it. After increasing the `bond_dim` from `2` to `4`, the error's gone. So I want to make sure if there is a good test scenario to deal with which circuit accepts `bond_dim=2` or not. could you please recommend it?
3. we can't use `ApplyFusedGate()` anymore. On top of it, `ApplyGate()` should be done with `main_circuits` not with `fused_circuits`.
4. Likewise, we can't use `ComputeExpectation`. We need to add similar function using `ApplyGate()`, not `ApplyFusedGate()`

Thank you so so much for your kind patience.

**Comment by jaeyoo** [comment]
Failed at (//tensorflow_quantum/core/ops/noise:noisy_samples_op_test )[https://github.com/tensorflow/quantum/pull/610/checks?check_run_id=3753494646#step:6:897] May i just re-run the test?

**Comment by jaeyoo** [comment]
PTAL

Note: `qsim::ApplyFusedGate` will be supported after `qsim==0.10.3` due to https://github.com/quantumlib/qsim/issues/395.

**Comment by MichaelBroughton** [comment]
We just cut a release here: https://github.com/quantumlib/qsim/releases/tag/v0.10.3-dev%2B20211001

Could we try upgrading to this new qsim version ?

**Comment by jaeyoo** [comment]
> We just cut a release here: https://github.com/quantumlib/qsim/releases/tag/v0.10.3-dev%2B20211001
> 
> Could we try upgrading to this new qsim version ?

Sure, it's my pleasure. Done.

File: tensorflow_quantum/core/ops/math_ops/simulate_mps_test.py [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+                symbol_names, symbol_values_array,
+                util.convert_to_tensor([[x] for x in pauli_sums]))
+
+        res = simulate_mps.mps_1d_expectation(
+            util.convert_to_tensor([cirq.Circuit() for _ in pauli_sums]),
+            symbol_names, symbol_values_array.astype(np.float64),
+            util.convert_to_tensor([[x] for x in pauli_sums]))
+        self.assertDTypeEqual(res, np.float32)
+
+    def test_simulate_mps_1d_expectation_results(self):
```
Comment (MichaelBroughton) [inline tensorflow_quantum/core/ops/math_ops/simulate_mps_test.py:244]: This is a good start. Could we add a few more tests here that do the following:

1. Test with all single qubit gates in `tfq.get_supported_gates`
2. Test with all single qubit gates in `tfq.get_supported_gates.controlled_by(a_qubit)` and ensure that this fails or passes accordingly)
3. Test that all two qubit gates in `tfq.get_supported_gates` works.
4. Ensure that the empty case works as expected

**Review by MichaelBroughton — CHANGES_REQUESTED** [review]
Looking great! Just a few moderate sized changes then we should be good to merge!

**Comment by MichaelBroughton** [comment]
I went ahead and cleaned up the implementation a little bit. Notable changes:

1. Removed the use of `ApplyFusedGate`. On closer inspection if two qubits that weren't adjacent to one another got fused together, things would break.
2. Added in the `swap_endian` option for parsing. The mps simulator in qsim has different endianness from the state vector simulator. Working in a proper fix to qsim would take a very long time so this is a workaround until we can get this fixed. https://github.com/quantumlib/qsim/issues/492
3. Removed the `ComputeLarge` since MPS has a much smaller memory footprint
4. Added a few more tests.

**Review by jaeyoo — COMMENTED** [review]
LGTM for me, but I can't approve my PR. :) @MichaelBroughton

**Pull request merged**